### PR TITLE
Improve `lj-fluid` validation and refactor code.

### PR DESCRIPTION
**Pull request by joaander**
## Description

* Provide ~10^(-5) level accuracy in `lj_fluid` cross validation test. `dt=0.0001` is required to do so.
* Plot NPT and NVT averages on separate lines.
* Add WCA validation test to `lj_fluid`.
* Allow `lj_fluid` to run on CPU only (with new config option `enable_gpu`).
* Refactor common plotting routines into `util.py`. 
* Include the simulation mode in the random number seed.
* Compute virial pressure in MC simulations.
* Plot quantity histograms averaged over all replicates.

## Motivation and context

* The previous test showed a clear separation between MC and MD results when run with 32 replicates.
* It seems impossible to get NVT and NPT to line up with ~10^(-5) accuracy, show averages separately as these are technically two different state points.
* The WCA tests are for comparison with MCCCS-MN.
* Earlier runs showed correlated errors across all Bussi thermostat runs. Incorporating the simulation mode into the RNG seed ensures each run within a job has a separate seed.

I will extend the use of the common plotting routines and other quality of life improvements to the other tests in future pull requests.

## How has this been tested?

Run on hoomd `trunk-major` 6c7eb5373` with 32 replicates:

lj_fluid_compare_kT1 0_density0 92_r_cut1 12_N1728
lj_fluid_compare_kT1 5_density0 6_r_cut2 5_N1728
lj_fluid_conservation_kT1 0_density0 92_r_cut1 12_N1728
lj_fluid_conservation_kT1 5_density0 6_r_cut2 5_N1728
lj_fluid_distribution_analyze_kT1 0_density0 92_r_cut1 12_N1728
lj_fluid_distribution_analyze_kT1 5_density0 6_r_cut2 5_N1728

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-validation/blob/trunk/CONTRIBUTING.md).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-validation/blob/trunk/ContributorAgreement.md).
- [x] My name is on the list of contributors (`CONTRIBUTORS.md`) in the pull request source branch.

### CODE REVIEW

File: hoomd_validation/util.py [hunk 1]
```diff
@@ -201,3 +202,154 @@ def make_simulation(
...
+
+def plot_distribution(ax, data, xlabel, expected=None, bins=100):
+    """Plot distributions."""
+    import numpy
+
+    max_density_histogram = 0
+    sim_modes = data.keys()
+
+    range_min = min([min(x) for x in data.values()])
+    range_max = max([max(x) for x in data.values()])
```
Comment (b-butler) [inline hoomd_validation/util.py:228]:
```suggestion
    range_min = min(min(x) for x in data.values())
    range_max = max(max(x) for x in data.values())
```
Use Python's generator syntax.

File: hoomd_validation/util.py [hunk 2]
```diff
@@ -201,3 +202,154 @@ def make_simulation(
...
+
+    for mode in sim_modes:
+        histogram, bin_edges = numpy.histogram(data[mode],
+                                               bins=bins,
+                                               range=(range_min, range_max),
+                                               density=True)
+
+        if numpy.all(numpy.asarray(data[mode]) == data[mode][0]):
+            histogram[:] = 0
```
Comment (b-butler) [inline hoomd_validation/util.py:237]:
```suggestion
        data_arr = np.asarray(data[mode])
        histogram, bin_edges = numpy.histogram(data_arr,
                                               bins=bins,
                                               range=(range_min, range_max),
                                               density=True)

        if numpy.all(data_arr == data_arr[0]):
            histogram[:] = 0
```
Might as well only convert once.

File: hoomd_validation/lj_fluid.py [hunk 3]
```diff
@@ -162,15 +187,15 @@ def make_md_simulation(job,
...
+    lj = md.pair.LJ(default_r_cut=job.statepoint.r_cut,
                     default_r_on=LJ_PARAMS['r_on'],
                     nlist=nlist)
     lj.params[('A', 'A')] = dict(sigma=LJ_PARAMS['sigma'],
                                  epsilon=LJ_PARAMS['epsilon'])
     lj.mode = 'xplor'
 
     # integrator
-    integrator = md.Integrator(dt=0.001, methods=[method], forces=[lj])
+    integrator = md.Integrator(dt=0.0001, methods=[method], forces=[lj])
```
Comment (b-butler) [inline hoomd_validation/lj_fluid.py:198]: Didn't you say 0.001 was sufficient?

File: hoomd_validation/lj_fluid.py [hunk 4]
```diff
@@ -938,178 +1015,192 @@ def lj_fluid_compare_modes(*jobs):
...
+
+            if 'md' in sim_mode:
+                ke = log_traj[
+                    'log/md/compute/ThermodynamicQuantities/kinetic_energy']
+                ke_means_expected[sim_mode].append(
+                    numpy.mean(ke) - 1 / 2 * n_dof * kT)
+                ke_sigmas_expected[sim_mode].append(
+                    numpy.std(ke) - 1 / math.sqrt(2) * math.sqrt(n_dof) * kT)
+
+                ke_samples[sim_mode].extend(list(ke))
```
Comment (b-butler) [inline hoomd_validation/lj_fluid.py:1137]: Why switch this array to a list?

File: hoomd_validation/lj_fluid.py [hunk 5]
```diff
@@ -938,178 +1015,192 @@ def lj_fluid_compare_modes(*jobs):
...
+            else:
+                pressure_samples[sim_mode].extend(
+                    list(log_traj['log/custom/virial_pressure']))
+
+            if 'nvt' in sim_mode:
+                density_samples[sim_mode].extend(
+                    list(log_traj['log/custom_actions/ComputeDensity/density']))
+            else:
+                density_samples[sim_mode].extend(
+                    list(log_traj['log/custom_actions/ComputeDensity/density']))
```
Comment (b-butler) [inline hoomd_validation/lj_fluid.py:1164]:
```suggestion
            density_samples[sim_mode].extend(
                list(log_traj['log/custom_actions/ComputeDensity/density']))
```

**Review by b-butler — COMMENTED** [review]
I have some suggestions. I am just providing a comment since I will be out next week, and won't be able to review then. The code looks good generally was a lot to review so I may have missed something.

File: hoomd_validation/lj_fluid.py (hunk 3 above)
Comment (joaander) [reply hoomd_validation/lj_fluid.py:198]: Yes, but I found that after making this pull request. Will rerun with new parameters.

File: hoomd_validation/lj_fluid.py (hunk 4 above)
Comment (joaander) [reply hoomd_validation/lj_fluid.py:1137]: I'm not sure. At one point, `ke_samples[sim_mode]` as a list of numpy arrays. I don't think this changed fixed that, but it was something I tried to fix it.

**Comment by joaander** [comment]
For the LJ potential, `r_cut=2.5` is the primary cause of the difference between the `rho` sampled by MC and MD. With `r_cut=4`, the difference is reduced significantly and all validation tests pass with `dt=0.001`.
Results with 87039a7a:
lj_fluid_compare_kT1 0_density0 92_r_cut1 12_N1728
lj_fluid_compare_kT1 5_density0 63_r_cut4 0_N4096
lj_fluid_conservation_kT1 0_density0 92_r_cut1 12_N1728
lj_fluid_conservation_kT1 5_density0 63_r_cut4 0_N4096
lj_fluid_distribution_analyze_kT1 0_density0 92_r_cut1 12_N1728
lj_fluid_distribution_analyze_kT1 5_density0 63_r_cut4 0_N4096

**Review by b-butler — APPROVED** [review]

**Pull request merged**